python/beeai_framework/agents/base.py

```diff
@@ -1,28 +1,85 @@
 # Copyright 2025 © BeeAI a Series of LF Projects, LLC
 # SPDX-License-Identifier: Apache-2.0
 
-from abc import ABC, abstractmethod
+from abc import abstractmethod
 from collections.abc import Awaitable, Callable
 from functools import cached_property
-from typing import Any, Generic, TypeVar
+from typing import Any, Unpack
 
 from pydantic import BaseModel
+from typing_extensions import TypeVar
 
 from beeai_framework.agents.errors import AgentError
 from beeai_framework.agents.types import AgentMeta
+from beeai_framework.backend import AnyMessage
 from beeai_framework.context import Run, RunContext, RunMiddlewareType
 from beeai_framework.emitter import Emitter
 from beeai_framework.memory import BaseMemory
+from beeai_framework.runnable import Runnable, RunnableOptions, RunnableOutput
 from beeai_framework.utils import AbortSignal
 
-TOutput = TypeVar("TOutput", bound=BaseModel)
 
+class AgentOptions(RunnableOptions, total=False):
+    """Agent options."""
 
-class BaseAgent(ABC, Generic[TOutput]):
-    def __init__(self) -> None:
-        super().__init__()
+    expected_output: str | type[BaseModel]
+    """
+    Instruction for steering the agent towards an expected output format.
+    This can be a Pydantic model for structured output decoding and validation.
+    """
+
+    total_max_retries: int
+    """
+    Maximum number of model retries.
+    """
+
+    max_retries_per_step: int
+    """
+    Maximum number of model retries per step.
+    """
+
+    max_iterations: int
+    """
+    Maximum number of iterations.
+    """
+
+    backstory: str
+    """
+    Additional piece of information or background for the agent.
+    """
+
+
+class AgentOutput(RunnableOutput):
+    """Agent output."""
+
+    output_structured: Any | BaseModel = None
+    """The formatted output returned by the agent."""
+
+
+R = TypeVar("R", default=AgentOutput, bound=AgentOutput)
+
+
+class BaseAgent(Runnable[R]):
+    """An abstract agent."""
+
+    def __init__(self, middlewares: list[RunMiddlewareType] | None = None) -> None:
+        super().__init__(middlewares=middlewares)
         self._is_running = False
-        self.middlewares: list[RunMiddlewareType] = []
+
+    @abstractmethod
+    def run(self, input: str | list[AnyMessage], /, **kwargs: Unpack[AgentOptions]) -> Run[R]:
+        """Execute an agent, specializing the runnable interface to accept a string as input
+        in addition to a list of messages.
+
+        Args:
+            input: The input to the agent
+            signal: The agent abort signal
+            context: A dictionary that can be used to pass additional context to the agent
+
+        Returns:
+            The agent output.
```

Comment:
I don't think this is needed since it is part of `AgentOptions` and visible to IDEs.

Reply:
This is in response to comments for other folks (Michael, Teryl, Anton, etc.)., that we should comment those kwargs well. So, let's err on the side of documenting them for this core interface, even though I understand the point.

Reply:
Sorry, could you rephrase the point? I understand commenting on the intent of the function, but not the kwargs in this specific case.